# Working with CSV Files

Welcome back! Let's discuss how to work with CSV files in Python. A file with the CSV file extension is a Comma Separated Values file. All CSV files are plain text, contain alphanumeric characters, and structure the data contained within them in a tabular form. Don't confuse Excel Files with csv files, while csv files are formatted very similarly to excel files, they don't have data types for their values, they are all strings with no font or color. They also don't have worksheets the way an excel file does. Python does have several libraries for working with Excel files, you can check them out [here](http://www.python-excel.org/) and [here](https://www.xlwings.org/).

Files in the CSV format are generally used to exchange data, usually when there's a large amount, between different applications. Database programs, analytical software, and other applications that store massive amounts of information (like contacts and customer data), will usually support the CSV format.

Let's explore how we can open a csv file with Python's built-in csv library. 

____
## Notebook Location. 

Run **pwd** inside a notebook cell to find out where your notebook is located

In [1]:
pwd

'C:\\Users\\KIIT\\Python Practice Files\\8-Advanced Python Modules\\Working with CSV Files'

## Opening A CSV Files

In [2]:
#The first step is to import the CSV library
import csv

In [3]:
#Second Step is opening the CSV file in the format <Variable>=open('<NameOfFile'.csv,mode='r')
initialise = open('example.csv',mode='r')

In [4]:
#Third step is to convert this CSV file into reader function in order to collect data and later function on it.
csvData=csv.reader(initialise)

In [5]:
#Fourth step is to convert this data into a readable format. For example say lists.
dataline=list(csvData)

UnicodeDecodeError: 'charmap' codec can't decode byte 0x8d in position 1810: character maps to <undefined>

As observable an encoding error is generated. This means our CSV file contains something that the CSV module cannot interpret using the ordinary encoding used for CSV files. For this we need to mentions specific encoding at the time of opening the CSV file in order to avoid such an error. Lets see how to do that and fix this error.

In [6]:
import csv

In [7]:
initialise=open('example.csv',encoding='utf-8')#Now python will use the UTF-8 encoding to decode this CSV file.

In [8]:
csvReader=csv.reader(initialise)

In [9]:
datalines=list(csvReader)

Note that CSV file is nothing but a file in which data is stored in a spreadsheet format(Tabular Format) which each data point seperated from the other by using comma in general.(There can be other things than comma). When we open this file in reader and pass this into a list the rows of the CSV file become elements of the main list and we obtain a 2-dimensional list. 

In [10]:
initialise.close()

## The with method
We have already discussed how we use the with method while we discussed working with file segment of python programming. Here is another example of implementation of with method alongside CSV file systems.

The general syntax is 

with open( < filename >.txt ) as < variable in which you want to save the file >:
    
    < Variable >=< Variable in which file is stored >.< any file method >
    
Examples will be seen below

In [11]:
with open('example.csv',encoding='utf-8') as withCSVfile:
    dataWithCSVFile=list(csv.reader(withCSVfile))

In [12]:
dataWithCSVFile

[['id', 'first_name', 'last_name', 'email', 'gender', 'ip_address', 'city'],
 ['1',
  'Joseph',
  'Zaniolini',
  'jzaniolini0@simplemachines.org',
  'Male',
  '163.168.68.132',
  'Pedro Leopoldo'],
 ['2',
  'Freida',
  'Drillingcourt',
  'fdrillingcourt1@umich.edu',
  'Female',
  '97.212.102.79',
  'Buri'],
 ['3',
  'Nanni',
  'Herity',
  'nherity2@statcounter.com',
  'Female',
  '145.151.178.98',
  'Claver'],
 ['4',
  'Orazio',
  'Frayling',
  'ofrayling3@economist.com',
  'Male',
  '25.199.143.143',
  'Kungur'],
 ['5',
  'Julianne',
  'Murrison',
  'jmurrison4@cbslocal.com',
  'Female',
  '10.186.243.144',
  'Sainte-Luce-sur-Loire'],
 ['6',
  'Lucy',
  'Gamet',
  'lgamet5@list-manage.com',
  'Female',
  '10.151.93.36',
  'Pak Phli'],
 ['7',
  'Dyana',
  'Howatt',
  'dhowatt6@amazon.com',
  'Female',
  '224.169.61.29',
  'Palmares'],
 ['8',
  'Kassey',
  'Herion',
  'kherion7@amazon.com',
  'Female',
  '245.51.154.79',
  'Zákynthos'],
 ['9',
  'Chrissy',
  'Hedworth',
  'chedworth8@ch

The advantage of using with is that is both open and closes the file automatically so we do no have to close the file manually

## The DictReader Class Of CSV library

In [13]:
with open('example.csv',encoding='utf-8') as withCSVfile:
    dataWithCSVFile=list(csv.DictReader(withCSVfile))

In [14]:
dataWithCSVFile

[{'id': '1',
  'first_name': 'Joseph',
  'last_name': 'Zaniolini',
  'email': 'jzaniolini0@simplemachines.org',
  'gender': 'Male',
  'ip_address': '163.168.68.132',
  'city': 'Pedro Leopoldo'},
 {'id': '2',
  'first_name': 'Freida',
  'last_name': 'Drillingcourt',
  'email': 'fdrillingcourt1@umich.edu',
  'gender': 'Female',
  'ip_address': '97.212.102.79',
  'city': 'Buri'},
 {'id': '3',
  'first_name': 'Nanni',
  'last_name': 'Herity',
  'email': 'nherity2@statcounter.com',
  'gender': 'Female',
  'ip_address': '145.151.178.98',
  'city': 'Claver'},
 {'id': '4',
  'first_name': 'Orazio',
  'last_name': 'Frayling',
  'email': 'ofrayling3@economist.com',
  'gender': 'Male',
  'ip_address': '25.199.143.143',
  'city': 'Kungur'},
 {'id': '5',
  'first_name': 'Julianne',
  'last_name': 'Murrison',
  'email': 'jmurrison4@cbslocal.com',
  'gender': 'Female',
  'ip_address': '10.186.243.144',
  'city': 'Sainte-Luce-sur-Loire'},
 {'id': '6',
  'first_name': 'Lucy',
  'last_name': 'Gamet',
  

Lets talk about it shall we? Here we have seen two different ways of opening a CSV file in python. They are:-

1)List of lists

2)List of dictionary

Now question is how are these different. To have clarity in understanding that, we need to go and see the different outputs we obtained for the two different approaches. We know that CSV file is tabular representation of data with the 1st row general being the header row followed by mulitple datasets with values in these datasets correspoding to columns created in header.

In the list of lists approach of opening a CSV file we observed that the header row was treated as a list and so were the datasets.

What changed in list of dictionary approach was that the columns in the header row were treated a keys in each row with corresponding values in datasets being treated as values in each and every row.

I hope you understood what I wanted to say.

## Reading A CSV File

In [15]:
datalines
#As observable we see that simple print the variable we stored the CSV file in the list format  

[['id', 'first_name', 'last_name', 'email', 'gender', 'ip_address', 'city'],
 ['1',
  'Joseph',
  'Zaniolini',
  'jzaniolini0@simplemachines.org',
  'Male',
  '163.168.68.132',
  'Pedro Leopoldo'],
 ['2',
  'Freida',
  'Drillingcourt',
  'fdrillingcourt1@umich.edu',
  'Female',
  '97.212.102.79',
  'Buri'],
 ['3',
  'Nanni',
  'Herity',
  'nherity2@statcounter.com',
  'Female',
  '145.151.178.98',
  'Claver'],
 ['4',
  'Orazio',
  'Frayling',
  'ofrayling3@economist.com',
  'Male',
  '25.199.143.143',
  'Kungur'],
 ['5',
  'Julianne',
  'Murrison',
  'jmurrison4@cbslocal.com',
  'Female',
  '10.186.243.144',
  'Sainte-Luce-sur-Loire'],
 ['6',
  'Lucy',
  'Gamet',
  'lgamet5@list-manage.com',
  'Female',
  '10.151.93.36',
  'Pak Phli'],
 ['7',
  'Dyana',
  'Howatt',
  'dhowatt6@amazon.com',
  'Female',
  '224.169.61.29',
  'Palmares'],
 ['8',
  'Kassey',
  'Herion',
  'kherion7@amazon.com',
  'Female',
  '245.51.154.79',
  'Zákynthos'],
 ['9',
  'Chrissy',
  'Hedworth',
  'chedworth8@ch

## Manipulation Of CSV Datafile Opened Via List of Lists Method

In [16]:
#Accessing the first row of the CSV file which by the way is now present as a list
datalines[0]

['id', 'first_name', 'last_name', 'email', 'gender', 'ip_address', 'city']

As observable we accessed the title file of the CSV file

In [17]:
#We can check the number of row present in a CSV file by checking the lenght of the list in which CSV file is saved.
len(datalines)

1001

As observable the CSV file 'example' has 1001 rows. Though it has 1000 datasets as the first row is the header row.

In [18]:
#We can also slice the CSV file list if we want
datalines[4:10]

[['4',
  'Orazio',
  'Frayling',
  'ofrayling3@economist.com',
  'Male',
  '25.199.143.143',
  'Kungur'],
 ['5',
  'Julianne',
  'Murrison',
  'jmurrison4@cbslocal.com',
  'Female',
  '10.186.243.144',
  'Sainte-Luce-sur-Loire'],
 ['6',
  'Lucy',
  'Gamet',
  'lgamet5@list-manage.com',
  'Female',
  '10.151.93.36',
  'Pak Phli'],
 ['7',
  'Dyana',
  'Howatt',
  'dhowatt6@amazon.com',
  'Female',
  '224.169.61.29',
  'Palmares'],
 ['8',
  'Kassey',
  'Herion',
  'kherion7@amazon.com',
  'Female',
  '245.51.154.79',
  'Zákynthos'],
 ['9',
  'Chrissy',
  'Hedworth',
  'chedworth8@china.com.cn',
  'Male',
  '124.222.93.57',
  'Boevange-sur-Attert']]

In [19]:
#We can also loop through the elements of CSV files saved in the form of lists.
for line in datalines[:5]:
    print(line)

['id', 'first_name', 'last_name', 'email', 'gender', 'ip_address', 'city']
['1', 'Joseph', 'Zaniolini', 'jzaniolini0@simplemachines.org', 'Male', '163.168.68.132', 'Pedro Leopoldo']
['2', 'Freida', 'Drillingcourt', 'fdrillingcourt1@umich.edu', 'Female', '97.212.102.79', 'Buri']
['3', 'Nanni', 'Herity', 'nherity2@statcounter.com', 'Female', '145.151.178.98', 'Claver']
['4', 'Orazio', 'Frayling', 'ofrayling3@economist.com', 'Male', '25.199.143.143', 'Kungur']


Let's imagine we wanted a list of  all the emails. For demonstration, since there are 1000 items plus the header, we will only do a few rows.

In [20]:
#We can also access specific elements of the CSV file saved in the form of lists.
#Lets say we want to access the email of a guy named Joseph(index 1).
datalines[1][3]

'jzaniolini0@simplemachines.org'

In [21]:
#Lets we want a list of all emails in our list.
allEmail=[]
for i in datalines:
    allEmail.append(i[3])

In [22]:
print(allEmail)

['email', 'jzaniolini0@simplemachines.org', 'fdrillingcourt1@umich.edu', 'nherity2@statcounter.com', 'ofrayling3@economist.com', 'jmurrison4@cbslocal.com', 'lgamet5@list-manage.com', 'dhowatt6@amazon.com', 'kherion7@amazon.com', 'chedworth8@china.com.cn', 'hgasquoine9@google.ru', 'ftarra@shareasale.com', 'abathb@umn.edu', 'lchastangc@goo.gl', 'cceried@yale.edu', 'ajeppe@umich.edu', 'pcaldeirof@chronoengine.com', 'nclunieg@utexas.edu', 'kroish@unicef.org', 'iboycotti@dailymail.co.uk', 'bglanertj@timesonline.co.uk', 'pcarolik@cam.ac.uk', 'vdranl@independent.co.uk', 'groantreem@nba.com', 'tdawidowitschn@free.fr', 'schoulertono@wisc.edu', 'eseawellp@live.com', 'daldrinq@wired.com', 'jdankr@miibeian.gov.cn', 'mlampels@a8.net', 'chalsteadt@amazon.de', 'gkarolowskiu@narod.ru', 'smundeev@nsw.gov.au', 'aoldershaww@so-net.ne.jp', 'bmattusevichx@foxnews.com', 'abohlensy@nih.gov', 'hbraidwoodz@amazon.com', 'hsudron10@nps.gov', 'jvanbrugh11@cocolog-nifty.com', 'hlightollers12@comsenz.com', 'egoodha

What if we wanted a list of full names?

In [23]:
full_names = []

for line in datalines:
    full_names.append(line[1]+' '+line[2])

In [24]:
full_names

['first_name last_name',
 'Joseph Zaniolini',
 'Freida Drillingcourt',
 'Nanni Herity',
 'Orazio Frayling',
 'Julianne Murrison',
 'Lucy Gamet',
 'Dyana Howatt',
 'Kassey Herion',
 'Chrissy Hedworth',
 'Hyatt Gasquoine',
 'Felicdad Tarr',
 'Andrew Bath',
 'Lucais Chastang',
 'Car Cerie',
 'Alvera Jepp',
 'Prescott Caldeiro',
 'Nariko Clunie',
 'Kalvin Rois',
 'Isa Boycott',
 'Benedetta Glanert',
 'Poul Caroli',
 'Virgie Dran',
 'Gillie Roantree',
 'Tabor Dawidowitsch',
 'Sarina Choulerton',
 'Elbert Seawell',
 'Danita Aldrin',
 'Jone Dank',
 'Merralee Lampel',
 'Charyl Halstead',
 'Giuseppe Karolowski',
 'Sutherlan Mundee',
 'Alexander Oldershaw',
 'Blayne Mattusevich',
 'Alfredo Bohlens',
 'Harman Braidwood',
 'Haslett Sudron',
 'Jeanelle VanBrugh',
 'Hetty Lightollers',
 'Elmira Goodhand',
 'Granger Lewerenz',
 'Tobie Lewson',
 'Knox Sainteau',
 'Eliot Vedekhov',
 'Honey Crenshaw',
 'Karmen Soppitt',
 'Fairleigh Sivess',
 'Angelina Stranio',
 'Lilla Erni',
 'Penn Batson',
 'Salvidor 

## Writing to CSV Files

We can also write csv files, either new ones or add on to existing ones.

### New File 
**This will also overwrite any exisiting file with the same name, so be careful with this!**

In [25]:
# newline controls how universal newlines works (it only applies to text
# mode). It can be None, '', '\n', '\r', and '\r\n'. ''

#The first step in writing a new CSV file is open the file or create a new file we want to write in.
#The general format of doing that is 
#<variable>=open('NameOfFile'.csv,mode='w',newline='')
file_to_output = open('to_save_file.csv',mode='w',newline='')

In [26]:
#Second step is to create a writer intance via writer attribute of csv library
#The general syntax used is
#<Variable>=csv.writer(<NameOfFileWithoutTheCSVExtension>,delimiter=',')
#Demiliter defines how each column will seperated from one another
csv_writer = csv.writer(file_to_output,delimiter=',')

In [27]:
csv_writer.writerow(['a','b','c'])#Declaring header of Csv file.
#Note how we use the writerow attribute of CSV library here.It is used along the variable where we created our writer 
#instance

7

In [28]:
csv_writer.writerows([['1','2','3'],['4','5','6']])#Here we are entering the datasets

In [29]:
file_to_output.close()#Here we are closing the CSV file.

____
### Existing File 

In [30]:
f = open('to_save_file.csv',mode='a',newline='')

In [31]:
csv_writer = csv.writer(f)#No delimiter used here as delimiter is already defined

In [32]:
csv_writer.writerow(['new','new','new'])

13

In [33]:
f.close()

That is all for the basics! If you believe you will be working with CSV files often, you may want to check out the powerful [pandas library](https://pandas.pydata.org/).